control-plane/subcommand/install-cni/kubeconfig.go

```diff
@@ -0,0 +1,146 @@
+package installcni
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"github.com/hashicorp/go-hclog"
+	"k8s.io/client-go/rest"
+)
+
+type KubeConfigFields struct {
+	KubernetesServiceProtocol string
+	KubernetesServiceHost     string
+	KubernetesServicePort     string
+	TLSConfig                 string
+	ServiceAccountToken       string
+}
+
+// createKubeConfig creates the kubeconfig file that the consul-cni plugin will use to communicate with the
+// kubernetes API.
+func createKubeConfig(mountedPath, kubeconfigFile string, logger hclog.Logger) error {
+
+	var kubecfg *rest.Config
+
+	// Get kube config information from cluster
+	kubecfg, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+	err = rest.LoadTLSFiles(kubecfg)
+	if err != nil {
+		return err
+	}
+
+	// Get the host, port and protocol used to talk to the kube API
+	kubeFields, err := KubernetesFields(kubecfg.CAData, logger)
+	if err != nil {
+		return err
+	}
+
+	// Write out the kubeconfig file
+	destFile := filepath.Join(mountedPath, kubeconfigFile)
+	err = writeKubeConfig(kubeFields, destFile, logger)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Wrote kubeconfig file", "name", destFile)
+	return nil
+}
+
+// KubernetesFields gets the needed fields from the in cluster config.
+func KubernetesFields(caData []byte, logger hclog.Logger) (*KubeConfigFields, error) {
+
+	var protocol = "https"
+	if val, ok := os.LookupEnv("KUBERNETES_SERVICE_PROTOCOL"); ok {
+		protocol = val
+	}
+
+	var serviceHost string
+	if val, ok := os.LookupEnv("KUBERNETES_SERVICE_HOST"); ok {
+		serviceHost = val
+	}
+
+	var servicePort string
+	if val, ok := os.LookupEnv("KUBERNETES_SERVICE_PORT"); ok {
+		servicePort = val
+	}
+
+	ca := "certificate-authority-data: " + base64.StdEncoding.EncodeToString(caData)
+
+	serviceToken, err := ServiceAccountToken()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debug("KubernetesFields: got fields", "protocol", protocol, "kubernetes host", serviceHost, "kubernetes port", servicePort)
+	return &KubeConfigFields{
+		KubernetesServiceProtocol: protocol,
+		KubernetesServiceHost:     serviceHost,
+		KubernetesServicePort:     servicePort,
+		TLSConfig:                 ca,
+		ServiceAccountToken:       serviceToken,
+	}, nil
+}
+
+// ServiceAccountToken gets the service token from a directory on the host.
+func ServiceAccountToken() (string, error) {
+	// serviceAccounttoken = /var/run/secrets/kubernetes.io/serviceaccount/token
+	token, err := ioutil.ReadFile(serviceAccountToken)
+	if err != nil {
+		return "", fmt.Errorf("could not read service account token: %v", err)
+	}
+	return string(token), nil
+
+}
+
+// writeKubeConfig writes out the kubeconfig file using a template.
+func writeKubeConfig(fields *KubeConfigFields, destFile string, logger hclog.Logger) error {
+
+	tmpl, err := template.New("kubeconfig").Parse(kubeconfigTmpl)
```

Comment:
I agree. I wanted to use a struct in the first place but I couldn't find a definition of the "Config" type in the kubeapi. I would rather use the official kubernetes type instead of redefining it myself. 

I will look into it.

Reply:
I think you might like this. I refactored the kubeconfig generation using the same Config structs that kubectl uses. I use the clientcmds Write function to generate the yaml output and it follows the same format that we have on ~/kube/config